Narrowing error on macOS

Compiling ops with fast_run on macOS results in an error because of "narrowing". This is a known issue (e.g. https://github.com/pymc-devs/pymc3/issues/3695) that is handled by downstream projects like PyMC3 (https://github.com/pymc-devs/pymc3/pull/3767). I propose that we fix this here by adding the `-Wno-c++11-narrowing` flag to `theano.config.gcc.cxxflags` by default on macOS.

**Please provide a minimal, self-contained, and reproducible example.**
```python
import theano
import theano.tensor as tt
a = tt.dmatrix()
b = tt.dvector()
c = tt.dmatrix()
theano.function([a, b, c], tt.dot(a, b + c))
```

**Please provide the full traceback of any errors.**
```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/compile/function.py", line 341, in function
    fn = pfunc(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/compile/pfunc.py", line 538, in pfunc
    return orig_function(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/compile/function_module.py", line 1997, in orig_function
    fn = m.create(defaults)
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/compile/function_module.py", line 1850, in create
    _fn, _i, _o = self.linker.make_thunk(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/link.py", line 738, in make_thunk
    return self.make_all(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/vm.py", line 1145, in make_all
    node.op.make_thunk(node, storage_map, compute_map, [], impl=impl)
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/op.py", line 980, in make_thunk
    return self.make_c_thunk(node, storage_map, compute_map, no_recycling)
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/op.py", line 877, in make_c_thunk
    outputs = cl.make_thunk(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cc.py", line 1275, in make_thunk
    cthunk, module, in_storage, out_storage, error_storage = self.__compile__(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cc.py", line 1204, in __compile__
    thunk, module = self.cthunk_factory(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cc.py", line 1720, in cthunk_factory
    module = get_module_cache().module_from_key(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cmodule.py", line 1229, in module_from_key
    module = lnk.compile_cmodule(location)
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cc.py", line 1610, in compile_cmodule
    module = c_compiler.compile_str(
  File "/Users/dforeman/src/pymc-devs/Theano-PyMC/theano/gof/cmodule.py", line 2567, in compile_str
    raise Exception(
Exception: ('The following error happened while compiling the node', Elemwise{add,no_inplace}(InplaceDimShuffle{x,0}.0, <TensorType(float64, matrix)>), '\n', "Compilation failed (return status=1): /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:524:27: error: non-constant-expression cannot be narrowed from type 'npy_intp' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing].     int init_totals[2] = {V5_n0, V3_n1};.                           ^~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:524:27: note: insert an explicit cast to silence this issue.     int init_totals[2] = {V5_n0, V3_n1};.                           ^~~~~.                           static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:524:34: error: non-constant-expression cannot be narrowed from type 'npy_intp' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing].     int init_totals[2] = {V5_n0, V3_n1};.                                  ^~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:524:34: note: insert an explicit cast to silence this issue.     int init_totals[2] = {V5_n0, V3_n1};.                                  ^~~~~.                                  static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:536:12: error: non-constant-expression cannot be narrowed from type 'ssize_t' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing].         0, V3_stride1, .            ^~~~~~~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:536:12: note: insert an explicit cast to silence this issue.         0, V3_stride1, .            ^~~~~~~~~~.            static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:537:1: error: non-constant-expression cannot be narrowed from type 'ssize_t' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing]. V5_stride0, V5_stride1, . ^~~~~~~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:537:1: note: insert an explicit cast to silence this issue. V5_stride0, V5_stride1, . ^~~~~~~~~~. static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:537:13: error: non-constant-expression cannot be narrowed from type 'ssize_t' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing]. V5_stride0, V5_stride1, .             ^~~~~~~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:537:13: note: insert an explicit cast to silence this issue. V5_stride0, V5_stride1, .             ^~~~~~~~~~.             static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:538:1: error: non-constant-expression cannot be narrowed from type 'ssize_t' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing]. V1_stride0, V1_stride1. ^~~~~~~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:538:1: note: insert an explicit cast to silence this issue. V1_stride0, V1_stride1. ^~~~~~~~~~. static_cast<int>( ). /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:538:13: error: non-constant-expression cannot be narrowed from type 'ssize_t' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing]. V1_stride0, V1_stride1.             ^~~~~~~~~~. /Users/dforeman/.theano/compiledir_macOS-10.15.7-x86_64-i386-64bit-i386-3.8.5-64/tmpsatm5hgs/mod.cpp:538:13: note: insert an explicit cast to silence this issue. V1_stride0, V1_stride1.             ^~~~~~~~~~.             static_cast<int>( ). 7 errors generated.. ", '[Elemwise{add,no_inplace}(<TensorType(float64, row)>, <TensorType(float64, matrix)>)]')
```

**Please provide any additional information below.**


## Versions and main components

* Theano version: GitHub master
* Theano config (`python -c "import theano; print(theano.config)"`)
* Python version: 3.8
* Operating system: macos
* How did you install Theano: pip (-e)

Hi just wondering if there is any updates on this?

I am running the following code:

```
import theano
import theano.tensor as tt

theano.config.gcc.cxxflags = "-Wno-c++11-narrowing"

a = tt.dmatrix()
b = tt.dvector()
c = tt.dmatrix()
theano.function([a, b, c], tt.dot(a, b + c))
```
And still receiving the same error:
```
Exception: ('The following error happened while compiling the node', Elemwise{add,no_inplace}(InplaceDimShuffle{x,0}.0, <TensorType(float64, matrix)>), '\n', "Compilation failed (return status=1): /Users/xxie/.theano/compiledir_Darwin-19.6.0-x86_64-i386-64bit-i386-3.7.9-64/tmprz_0y49f/mod.cpp:524:27: error: non-constant-expression cannot be narrowed from type 'npy_intp' (aka 'long') to 'int' in initializer list [-Wc++11-narrowing].     int init_totals[2] = {V3_n0, V5_n1};.                           ^~~~~. /Users/xxie/.theano/compiledir_Darwin-19.6.0-x86_64-i386-64bit-i386-3.7.9-64/tmprz_0y49f/mod.cpp:524:27: note: insert an explicit cast to silence this issue.
```

My settings on MacOS Catalina:
```
python --version: Python 3.7.9

conda list | grep theano
theano                    1.0.5            py37h54c7649_1    conda-forge
theano-pymc               1.1.0            py37h54c7649_1    conda-forge
```

My end goal is to use `theano` or `aesara`'s `rfft`, but obviously that function compile is failing with the same issue. Please let me know if there's any progress or if i can provide more info. Thanks in advance.

--- a/environment-arm.yml
+++ b/environment-arm.yml
@@ -1,0 +1,47 @@
+# To use:
+#
+#   $ conda env create -f environment.yml  # `mamba` works too for this command
+#   $ conda activate aesara-dev
+#
+name: aesara-dev
+channels:
+  - conda-forge
+dependencies:
+  - python
+  - compilers
+  - numpy>=1.17.0
+  - scipy>=0.14
+  - filelock
+  - etuples
+  - logical-unification
+  - miniKanren
+  - cons
+  # Non-Intel BLAS
+  - nomkl
+  - openblas
+  - libblas=*=*openblas
+  # numba backend
+  - numba>=0.55.2
+  - llvmlite>=0.38.1
+  - numba-scipy
+  # For testing
+  - coveralls
+  - diff-cover
+  - pytest
+  - pytest-cov
+  - pytest-xdist
+  # For building docs
+  - sphinx>=1.3
+  - sphinx_rtd_theme
+  - pygments
+  - pydot
+  - ipython
+  # developer tools
+  - pre-commit
+  - packaging
+  - typing_extensions
+  # optional
+  - sympy
+  - cython
+  - jax
+  - jaxlib

